### DISCUSSION TITLE
Storage for scanlist assignments

### DISCUSSION THREAD

**Discussion by hdavid** [AubsUK/uv-k5-firmware-custom#20] [Ideas]
Hi,

(this is more a request for chat than an actual issue)

your fork and the idea of the multiple scanlist, is great! 
 
I wanted to take a look and see if i could merge your idea with some of the UI features from https://github.com/joaquimorg/uv-k5-firmware-custom, but then realised they would not fit in the flash...

Anyhow i looked at the code, and i was wonderning why it is allowing to assign multiple scanlist to a channel, as we can scan for multiple scan list. Assigning a single number of each channel would reduce the storage required. 
if i got it right, two bytes are used per memory channel. then with this change, we could use just one byte for each channel, saving about 200bytes flash.

Also, maybe the 10 scanlist could be a feature flag, merged back to egzummer and also merged into the uv-k5 driver included in chirp. this would make this feature so much more visible and give it the visibility it deserves ! i already prepared some code on a branch for this if you wish.

regards,

Henri

**Comment by AubsUK** [comment]
Thanks for the feedback. I'll take a look at the repo you provided and see what it's like.

The ScanaLists for each channel takes up 10 bits plus 1 bit for lockout, so yes, 2 bytes. It uses memory locations that were originally used for the 10 character Channel Name (16 bytes each but only consumes 10 bytes), so was not previously used for anything else, and removing it wouldn't give any additional space for compiled code as I understand that's stored separately.

**Comment by hdavid** [comment]
Thanks for the explanation. i should have mentioned that i was hoping that the recovered eeprom could be used to renables these bands https://github.com/AubsUK/uv-k5-firmware-custom/blob/main/settings.c#L33 , to keep the feature parity with egzumer, but then again, they might be stored in a different location.
Anyhow, your firmware flavour is now my daily driver :)

**Comment by AubsUK** [comment]
(Moved from [Issues](https://github.com/AubsUK/uv-k5-firmware-custom/issues) to [Discussions](https://github.com/AubsUK/uv-k5-firmware-custom/discussions/))

They aren't bands, they're just the default pre-programmed memory channels that are created when you factory reset the radio.  Previously it would create 5 default channels, I removed all but the first (145.0 MHz) to save some space in the compiled code.  Other than the additions and minor tweaks I've made, the code is pretty much identical to the original from Egzumer.

They're called from `SETTINGS_FactoryReset()` on [Line447](https://github.com/AubsUK/uv-k5-firmware-custom/blob/main/settings.c#L447), and specifically referenced in [Line497](https://github.com/AubsUK/uv-k5-firmware-custom/blob/main/settings.c#L497).

Thank you, it's pleasing to hear someone else likes what I've done!  Please keep the ideas coming, but I can't guarantee to do them all.  I had a look at the other firmware, I'm not sure what's special about it that makes it worthwhile to add some of the features.

It does work with CHIRP, but you need to use my [forked driver](https://github.com/AubsUK/uvk5-chirp-driver), I trust you've seen that for easy editing?

For reference, the idea I had about channels being able to go into multiple ScanLists was primarily for when moving around to different locations, especially work, home, commuting, regular travel, etc. but thinking about it, I probably won't use it myself.  There's only a few people so far that I've seen using my firmware, so it would be good to get their opinion if they use it or not - I could probably save a bit of extra complied code space.

The firmware build shows `Firmware uses 99.04% of available memory (60848/61439 bytes)` - 591 bytes free at the moment, but I have plans to look into utilising the mechanism the Spectrum Analyser uses into a Channel Sweep to speed up the scanning of memories (whether or not it'll materialise, who knows!).

My current ScanList usage is as laid out below, it makes it easy to remember what's for what and quick to turn on/off:

<table style="border-spacing:30px; border: 1px solid black;">
<tr>
<td align="center"><b>1</b><br />UHF Repeaters<br />(local to home)</td>
<td align="center"><b>2</b><br />VHF Repeaters<br />(local to home)</td>
<td align="center"><b>3</b><br />PMR<br />&nbsp;</td>
<td align="center"><b>*</b><br />&nbsp;<br />&nbsp;</td>
</tr>
<tr>
<td align="center"><b>4</b><br />UHF<br />Channels</td>
<td align="center"><b>5</b><br />VHF<br />Channels</td>
<td align="center"><b>6</b><br />Marine<br />&nbsp;</td>
<td align="center"><b>0</b><br />unused<br />&nbsp;</td>
</tr>
<tr>
<td align="center"><b>7</b><br />CB UK<br />40 FM</td>
<td align="center"><b>8</b><br />CB CEPT<br />40 FM</td>
<td align="center"><b>9</b><br />unused<br />&nbsp;</td>
<td align="center"><b>F</b><br />&nbsp;<br />&nbsp;</td>
</tr>
</table>
I could use 0 and 9 for when I travel I suppose.


I've never heard anything locally on the CB frequencies (I do have a 27MHz rubber duck and telescopic antennas), so I might remove them if I have something else worthwhile to go in them.

I'm loving the * to temporarily skip noisy channels!
